src/provider/meta.test.ts

```diff
@@ -0,0 +1,70 @@
+import { ExternalProvider, JsonRpcProvider } from '@ethersproject/providers'
+import WalletConnectProvider from '@walletconnect/ethereum-provider'
+
+import { getWalletMeta, getWalletName, WalletMeta, WalletType } from './meta'
+
+class MockJsonRpcProvider extends JsonRpcProvider {
+  name = WalletType.UNKNOWN
+  arg: string
+
+  constructor(arg?: unknown) {
+    super()
+    this.arg = JSON.stringify(arg)
+  }
+}
+
+class MockWalletConnectProvider extends MockJsonRpcProvider {
+  name = WalletType.WALLET_CONNECT
+  provider: ExternalProvider
+
+  constructor(peerMeta: WalletConnectProvider['connector']['peerMeta']) {
+    super(peerMeta)
+    this.provider = { isWalletConnect: true, connector: { peerMeta } } as ExternalProvider
+  }
+}
+
+class MockInjectedProvider extends MockJsonRpcProvider {
+  name = WalletType.INJECTED
+  provider: ExternalProvider
+
+  constructor(provider: Record<string, boolean | undefined>) {
+    super(provider)
+    this.provider = {
+      isConnected() {
+        return true
+      },
+      ...provider,
+    } as ExternalProvider
+  }
+}
+
+const PEER_META = { name: 'name', description: 'description', url: 'url', icons: [] }
+
+const testCases: [MockJsonRpcProvider, WalletMeta][] = [
+  [new MockJsonRpcProvider(), { type: WalletType.UNKNOWN }],
+  [new MockWalletConnectProvider(null), { type: WalletType.WALLET_CONNECT }],
+  [new MockWalletConnectProvider(PEER_META), { type: WalletType.WALLET_CONNECT, ...PEER_META }],
+  [new MockInjectedProvider({}), { type: WalletType.INJECTED, name: '' }],
+  [new MockInjectedProvider({ isMetaMask: false }), { type: WalletType.INJECTED, name: '' }],
+  [new MockInjectedProvider({ isMetaMask: true }), { type: WalletType.INJECTED, name: 'MetaMask' }],
+  [new MockInjectedProvider({ isTest: true, isMetaMask: true }), { type: WalletType.INJECTED, name: 'Test MetaMask' }],
```

Comment:
Now that we have this test suite in place, another side comment, we should add a follow-up action item to move `isMetaMask`, `isCoinbase` etc. here, instead of living in the `interface`.